src/core/namespace.c

```diff
@@ -3310,6 +3313,379 @@ bool ns_type_supported(NamespaceType type) {
         return access(ns_proc, F_OK) == 0;
 }
 
+static int unpeel_get_fd(const char *mount_path, int *ret_fd) {
+        _cleanup_close_pair_ int pipe_fds[2] = EBADF_PAIR;
+        pid_t pid;
+        int r;
+
+        assert(mount_path);
+        assert(ret_fd);
+
+        r = socketpair(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC, 0, pipe_fds);
+        if (r < 0)
+                return log_debug_errno(errno, "Failed to create socket pair: %m");
+
+        /* Clone mount namespace here to unpeel without affecting live process */
+        r = safe_fork("(sd-ns-unpeel)", FORK_DEATHSIG_SIGTERM|FORK_LOG|FORK_WAIT|FORK_NEW_MOUNTNS|FORK_MOUNTNS_SLAVE, &pid);
+        if (r < 0)
+                return r;
+        if (r == 0) {
+                _cleanup_close_ int dir_fd = -EBADF;
+
+                pipe_fds[0] = safe_close(pipe_fds[0]);
+
+                /* Opportunistically unmount any overlay at this path */
+                if (path_is_fs_type(mount_path, OVERLAYFS_SUPER_MAGIC)) {
+                        r = umount_recursive(mount_path, MNT_DETACH);
+                        if (r == 0) { /* no umounts done, possible if a previous reload deleted all extensions */
+                                log_debug("No overlay layer unmountable from %s", mount_path);
+                        } else if (r < 0) {
+                                log_debug_errno(r, "Failed to unmount %s: %m", mount_path);
+                                _exit(EXIT_FAILURE);
+                        }
+                }
+
+                /* Now that /mount_path is exposed, get an FD for it and pass back */
+                dir_fd = open_tree(-EBADF, mount_path, AT_SYMLINK_NOFOLLOW|OPEN_TREE_CLONE);
+                if (dir_fd < 0) {
+                        log_debug_errno(errno, "Failed to clone mount %s: %m", mount_path);
+                        _exit(EXIT_FAILURE);
+                }
+
+                r = send_one_fd(pipe_fds[1], dir_fd, 0);
+                if (r < 0) {
+                        log_debug_errno(r, "Failed to send mount fd: %m");
+                        _exit(EXIT_FAILURE);
+                }
+
+                _exit(EXIT_SUCCESS);
+        }
+
+        pipe_fds[1] = safe_close(pipe_fds[1]);
+
+        r = receive_one_fd(pipe_fds[0], 0);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to receive mount fd: %m");
+
```

Comment:
it might be nice to have a safety check here that validates that what we acquired here is not in fact an overlayfs itself, because then things would be really weird. i.e. add a simple check like this here right after opening:

```c
r = fd_is_fs_type(unpeeled_fd, OVERLAYFS_SUPER_MAGIC);
if (r < 0)
        return log_debug_errno(r, "Unable to determine if unpeeled directory refers to overlayfs: %m");
if (r)
        return log_debug_errno(SYNTHETIC_ERRNO(ENORECOVERY), "Unpeeled mount is still an overlayfs, something is weird, refusing.");
``